src/openapi/lab.yml

```diff
@@ -860,18 +956,148 @@ endpoints:
+          application/json:
+            schema:
+              $ref: '#/model/Cluster'
       default:
         $ref: 'common.yml#/responses/problem'
+    security:
+      - oauth2: []
 
   cluster_delete:
     summary: Delete cluster
     tags: [lab]
     operationId: rhub.api.lab.cluster.delete_cluster
+    parameters:
+      - $ref: '#/parameters/cluster_id'
+    responses:
+      '204':
+        description: Success
+      default:
+        $ref: 'common.yml#/responses/problem'
+    security:
+      - oauth2: []
+
+  cluster_list_events:
+    summary: Get cluster events list
+    tags: [lab]
+    operationId: rhub.api.lab.cluster.list_cluster_events
     parameters:
       - $ref: '#/parameters/cluster_id'
     responses:
       '200':
+        description: Cluster events
+        content:
+          application/json:
+            schema:
+              type: array
+              items:
+                $ref: '#/model/ClusterEvent'
+      default:
+        $ref: 'common.yml#/responses/problem'
+    security:
+      - oauth2: []
+
+  cluster_event_get:
+    summary: Get cluster event
+    tags: [lab]
+    operationId: rhub.api.lab.cluster.get_cluster_event
+    parameters:
+      - $ref: '#/parameters/event_id'
+    responses:
+      '200':
+        description: Cluster event
+        content:
+          application/json:
+            schema:
+              $ref: '#/model/ClusterEvent'
+      default:
+        $ref: 'common.yml#/responses/problem'
+    security:
+      - oauth2: []
+
+  cluster_event_get_stdout:
+    summary: Get cluster event output
+    tags: [lab]
+    operationId: rhub.api.lab.cluster.get_cluster_event_stdout
+    parameters:
+      - $ref: '#/parameters/event_id'
+    responses:
+      '200':
+        description: Event output
+        content:
+          text/plain:
+            schema:
+              type: string
+      default:
+        $ref: 'common.yml#/responses/problem'
+    security:
+      - oauth2: []
+
+  cluster_hosts_list:
+    summary: Get cluster hosts
+    tags: [lab]
+    operationId: rhub.api.lab.cluster.list_cluster_hosts
+    parameters:
+      - $ref: '#/parameters/cluster_id'
+    responses:
+      '200':
+        description: Cluster hosts
+        content:
+          text/plain:
+            schema:
+              type: array
+              items:
+                $ref: '#/model/ClusterHost'
+      default:
+        $ref: 'common.yml#/responses/problem'
+    security:
+      - oauth2: []
+
+  cluster_hosts_create:
+    summary: Create or update cluster hosts
+    tags: [lab]
+    operationId: rhub.api.lab.cluster.create_cluster_hosts
+    parameters:
+      - $ref: '#/parameters/cluster_id'
+    requestBody:
+      required: true
+      content:
+        application/json:
+          schema:
+            type: array
+            items:
+              allOf:
+                - $ref: '#/model/ClusterHost'
+                - required:
+                    - fqdn
+                    - ipaddr
```

Comment:
if cluster_hosts are to simulate the hosts table currently in QL, the purpose is to track resource usage for each host in the cluster so they can be summed up both for the cluster as well as for overall region usage then possibly the additional fields related to resource usage should be required such as what is in the hosts table below:

MariaDB [quicklab]> describe hosts;
+-------------+-----------------------+------+-----+---------+----------------+
| Field       | Type                  | Null | Key | Default | Extra          |
+-------------+-----------------------+------+-----+---------+----------------+
| id          | int(11) unsigned      | NO   | PRI | NULL    | auto_increment |
| cluster_id  | int(11)               | NO   |     | NULL    |                |
| fqdn        | varchar(255)          | YES  |     | NULL    |                |
| ipv4addr    | varchar(16)           | YES  |     | NULL    |                |
| num_vcpus   | tinyint(3) unsigned   | YES  |     | NULL    |                |
| ram_mb      | mediumint(8) unsigned | YES  |     | NULL    |                |
| num_volumes | tinyint(3) unsigned   | YES  |     | NULL    |                |
| volumes_gb  | smallint(5) unsigned  | YES  |     | NULL    |                |
+-------------+-----------------------+------+-----+---------+----------------+

Reply:
No, it is not just to simulate QL tables. It may be useful to get the hosts list via API, that's why I added this. We can use something like collectd to get resource usage,. that will give us more useful data than a simple table.

Reply:
Ok.  If collectd or some other tool is used to collect resource usage that's fine.  It will need to be fast both to collect the resource usage data for an individual (or possibly team) as well as overall usage for an entire "region" to prevent overuse.  QL currently has such a restriction in place and highly probable that RHub would need something similar so a decision to allow/disallow a new cluster can be made nearly instantaneously.